### PR TITLE
Implement hdt_get_function_pointer

### LINKED ISSUE #35465: Support custom hosting a hybrid between app and component
In .NET Core 3.x, work was done to make custom hosting easier. Support was added for an "application" and a "component". I need support for a hybrid between these two.

The general use case is for an application where its features are spread across native and managed code, with the entry point being in native code. The details of my specific use case are here: https://github.com/dotnet/runtime/issues/34946.

It's like an "application" but not "component":
* This is the main application.
  -Missing dependencies are an error.
  -Loading into default ALC is expected.
  -Any isolation must be implemented by the app itself.
* It may be deployed as self-contained or framework-dependent.

It's like a "component" but not an "application":
* No `static void Main(string[] args)`, no concept of args
* Entry point is custom
* Managed code needs to continue running after execution has passed back to native code

Unique aspects for this hybrid scenario:
* The native code portion will initialize the .NET Core runtime very early in the lifetime of the process.
* If deployed as self-contained, then it was published with a RID compatible with the native code portion of the application.

Reasons to be implemented in .NET Core instead of by the custom host using coreclr directly:
* Logic to find the framework requested by the app is very complex and is subject to change with each new release of the framework.
* Building the TPA list from deps.json files is very complex and is subject to change with each new release of the framework.
* Various documentation discourages using coreclr directly, and some even make it seem like it could be deprecated in the future.

## Proposal
<details>
    <summary>Original proposal</summary>
Add two new exported functions to `hostfxr` - `hostfxr_initialize_for_managed_host` and `hostfxr_create_delegate`. These two functions are similar to the existing `hostfxr_initialize_for_dotnet_command_line`/`hostfxr_run_app` and `hostfxr_initialize_for_runtime_config`/`hostfxr_get_runtime_delegate` function pairs in that the handle created from `hostfxr_initialize_for_managed_host` can only be used with `hostfxr_create_delegate`.
    
    //
    // Initializes the hosting components using an application
    //
    // Parameters:
    //    managed_host_path
    //      Path to the managed host assembly file
    //    deps_json_path
    //      Optional. Path to the .deps.json file
    //    runtime_config_path
    //      Optional. Path to the .runtimeconfig.json file
    //    parameters
    //      Optional. Additional parameters for initialization
    //    host_context_handle
    //      On success, this will be populated with an opaque value representing the initialized host context
    //
    // Return value:
    //    Success          - Hosting components were successfully initialized
    //    HostInvalidState - Hosting components are already initialized
    //
    // This function will find the .runtimeconfig.json and .deps.json corresponding to the given 
         application
    // with which to resolve frameworks and dependencies and prepare everything needed to load the runtime.
    //
    // This function does not load the runtime.
    //
    typedef int32_t(HOSTFXR_CALLTYPE *hostfxr_initialize_for_managed_host_fn)(
        const char_t *managed_host_path,
        const char_t *deps_json_path,
        const char_t *runtime_config_path,
        const struct hostfxr_initialize_parameters *parameters,
        /*out*/ hostfxr_handle *host_context_handle);
    
    //
    // Create a native callable function pointer for a managed method from the currently loaded CoreCLR or from a newly created one.
    //
    // Parameters:
    //     host_context_handle
    //       Handle to the initialized host context
    //     entry_point_assembly_name
    //       Name of the assembly which holds the custom entry point
    //     entry_point_type_name
    //       Name of the type which holds the custom entry point
    //     entry_point_method_name
    //       Name of the method which is the custom entry point
    //     delegate
    //       Output parameter, the function stores a native callable function pointer to the delegate at the specified address
    //
    // Returns:
    //     The error code result.
    //
    // The host_context_handle must have been initialized using hostfxr_initialize_for_managed_host.
    // The assembly is loaded into the default AssemblyLoadContext.
    //
    typedef int32_t(HOSTFXR_CALLTYPE *hostfxr_create_delegate_fn)(
        const hostfxr_handle host_context_handle,
        const char_t *entry_point_assembly_name,
        const char_t *entry_point_type_name,
        const char_t *entry_point_method_name,
        /*out*/ void **delegate);

`hostfxr_initialize_for_managed_host` is essentially the same as `hostfxr_initialize_for_dotnet_command_line` but instead of `argc`/`argv`, it takes the application, deps.json, and runtimeconfig.json as explicit parameters.

`hostfxr_create_delegate` is simply a wrapper around the create delegate functionality of coreclr. `hostfxr_get_runtime_delegate` could not be reused since it always loads the target assembly into an isolated ALC (which is why it requires a handle from `hostfxr_initialize_for_runtime_config` - isolation is not supported in self-contained applications and it blocks on SCD).
</details>

Design document update and discussion: https://github.com/dotnet/runtime/pull/36990

### PR DESCRIPTION
**Pull request by rseanhall**
Second half of https://github.com/dotnet/runtime/issues/35465. First half was at https://github.com/dotnet/runtime/pull/37473.

Closes https://github.com/dotnet/runtime/issues/35465.

### CODE REVIEW

File: src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/FunctionPointerLoader.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        /// Native hosting entry point for creating a native delegate
+        /// </summary>
+        /// <param name="typeNameNative">Assembly qualified type name</param>
+        /// <param name="methodNameNative">Public static method name compatible with delegateType</param>
+        /// <param name="delegateTypeNative">Assembly qualified delegate type name</param>
+        /// <param name="loadContext">Extensibility parameter (currently unused)</param>
+        /// <param name="reserved">Extensibility parameter (currently unused)</param>
+        /// <param name="functionHandle">Pointer where to store the function pointer result</param>
+        [UnmanagedCallersOnly]
+        public static unsafe int GetFunctionPointer(IntPtr typeNameNative,
```
Comment (vitek-karas) [inline src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/FunctionPointerLoader.cs:49]: I would add this to the `ComponentActivator` and not create a new class:
* No need to duplicate code (the `MarshalToString` method for example)
* We can share the `s_delegates`, no reason to have two dictionaries for basically the same thing
* More options for sharing code - the method body here is very similar to the `LoadAssemblyAndGetFunctionPointer` so ideally we would rewrite it such that there's little to no duplication

Comment (vitek-karas) [reply src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/FunctionPointerLoader.cs:49]: I do see that the name `ComponentActivator` will then become somewhat weird. We can discuss changing that if we feel it's really bad. Even though it's a public class, it's not part of any public contract - `System.Private.CoreLib` has quite a few public APIs which are not truly public (nobody compiles directly against S.P.CoreLib except the rest of the framework). And the only other place where it's used is `hostpolicy` which always ships with the runtime, so there will never be a case where older/newer `hostpolicy` would interact with newer/older runtime.

File: src/installer/corehost/cli/test/nativehost/host_context_test.cpp [hunk 2]
```diff
@@ -380,6 +420,91 @@ namespace
 
         return rc == StatusCode::Success && rcClose == StatusCode::Success;
     }
+
+    bool get_function_pointer_test(
```
Comment (vitek-karas) [inline src/installer/corehost/cli/test/nativehost/host_context_test.cpp:424]: I would rename this to something like `component_get_function_pointer_test` to make it clear that this is on the "component" context.

File: src/installer/corehost/cli/test/nativehost/host_context_test.cpp [hunk 3]
```diff
@@ -625,3 +750,26 @@ bool host_context_test::app_load_assembly_and_get_function_pointer(
 
     return app_load_assembly_and_get_function_pointer_test(hostfxr, argc, argv, config_log_prefix, test_output);
 }
+
+bool host_context_test::get_function_pointer(
```
Comment (vitek-karas) [inline src/installer/corehost/cli/test/nativehost/host_context_test.cpp:754]: `component_get_function_pointer`

File: src/installer/tests/HostActivation.Tests/NativeHosting/FunctionPointerLoading.cs [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+using Microsoft.DotNet.Cli.Build.Framework;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Xunit;
+
+namespace Microsoft.DotNet.CoreSetup.Test.HostActivation.NativeHosting
+{
+    public partial class FunctionPointerLoading : IClassFixture<ComponentActivation.SharedTestState>
```
Comment (vitek-karas) [inline src/installer/tests/HostActivation.Tests/NativeHosting/FunctionPointerLoading.cs:13]: I would:
* Rename the existing `ComponentActivation` test to `LoadAssemblyAndGetFunctionPointer`
* Rename this to `GetFunctionPointer`

To make it much more obvious which test belongs to what runtime functionality.

File: src/installer/tests/HostActivation.Tests/NativeHosting/FunctionPointerLoading.cs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        public FunctionPointerLoading(ComponentActivation.SharedTestState sharedTestState)
+        {
+            sharedState = sharedTestState;
+        }
+
+        [Theory]
+        [InlineData(true, true)]
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        public void CallDelegate(bool validType, bool validMethod)
```
Comment (vitek-karas) [inline src/installer/tests/HostActivation.Tests/NativeHosting/FunctionPointerLoading.cs:29]: Please be explicit which context - this one means "app" implicitly, in the product you use "component" implicitly... I personally don't mind long descriptive names, so `CallDelegateOnApplicationContext` for example...

File: src/installer/tests/HostActivation.Tests/NativeHosting/FunctionPointerLoading.cs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.DotNet.Cli.Build.Framework;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Xunit;
+
+namespace Microsoft.DotNet.CoreSetup.Test.HostActivation.NativeHosting
```
Comment (vitek-karas) [inline src/installer/tests/HostActivation.Tests/NativeHosting/FunctionPointerLoading.cs:11]: The main scenario for these changes is to support `get_function_pointer` on a self-contained "app/component", can you please add tests for that as well?

**Review by vitek-karas — CHANGES_REQUESTED** [review]

File: src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/FunctionPointerLoader.cs (hunk 1 above)
Comment (rseanhall) [reply src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/FunctionPointerLoader.cs:49]: My thinking was that the assembly path parameter made it fundamentally different, along with the fact that the future extensibility parameters were also likely to be different from each other. There might be a problem with the design - how is `get_function_pointer` ever going to support an isolated ALC if it doesn't have a way to specify the assembly path?

Comment (vitek-karas) [reply src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/FunctionPointerLoader.cs:49]: My view is that `LoadAssemblyAndGetFunctionPointer` is basically a concatenation of two relatively distinct parts:
* Load assembly
* Get function pointer

You're now adding the second part as a separately accessible piece (with limitations). The goal of "Get function pointer" should not be to worry about loading assemblies or creating load contexts. That's the job of the first part "Load assembly". Long term I would like to introduce a separately callable function which does just that part - load assembly with some options as to where (default ALC, new isolated ALC, existing secondary ALC, ...). That's why we're adding the `loadContext` parameter to `get_function_pointer` - basically the future "load assembly" would return the value to be passed to `get_function_pointer`.

I don't see a problem with the specific runtime helpers (`LoadAssemblyAndGetFunctionPointer` and now `GetFunctionPointer`) having different parameters and somewhat different behavior while living on the same class. To me they have a lot more in common as can be seen by the implementation which is very similar.

If you think that having a separate class would be better, I don't have a problem with that (I personally don't see the point, but that's just me) - what I do want to see is sharing the code which should work the same.

**Comment by rseanhall** [comment]
If you don't mind, I'd like to address your feedback on the existing code first in https://github.com/dotnet/runtime/pull/37784 before continuing with this PR.

File: src/installer/tests/HostActivation.Tests/NativeHosting/GetFunctionPointer.cs [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+                validType ? sharedState.ComponentTypeName : $"Component.BadType, {componentProject.AssemblyName}",
+                validMethod ? sharedState.ComponentEntryPoint1 : "BadMethod",
+            };
+            CommandResult result = sharedState.CreateNativeHostCommand(args, sharedState.DotNetRoot)
+                .Execute();
+
+            result.Should()
+                .InitializeContextForConfig(componentProject.RuntimeConfigJson);
+
+            result.Should().Fail();
```
Comment (vitek-karas) [inline src/installer/tests/HostActivation.Tests/NativeHosting/GetFunctionPointer.cs:79]: Please add a comment why it always fails (the component type won't resolve in the default ALC)

**Review by vitek-karas — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by vitek-karas** [comment]
Thanks a lot @rseanhall!

**Comment by nxrighthere** [comment]
Oh, this is just great. I wish we had this in .NET Core 3, my [project](https://github.com/nxrighthere/UnrealCLR) would benefit a lot from it.

**Pull request merged**